python/plugins/processing/tools/postgis.py

```diff
@@ -138,41 +172,73 @@ def _quote(self, ident):
 
 class GeoDB:
 
+    @classmethod
+    def from_name(cls, conn_name):
+        uri = uri_from_name(conn_name)
+        return cls(uri=uri)
+
     def __init__(self, host=None, port=None, dbname=None, user=None,
-                 passwd=None):
+                 passwd=None, service=None, uri=None):
         # Regular expression for identifiers without need to quote them
         self.re_ident_ok = re.compile(r"^\w+$")
 
-        self.host = host
-        self.port = port
-        self.dbname = dbname
-        self.user = user
-        self.passwd = passwd
-
-        if self.dbname == '' or self.dbname is None:
-            self.dbname = self.user
-
-        try:
-            self.con = psycopg2.connect(self.con_info())
-        except psycopg2.OperationalError as e:
-            raise DbError(unicode(e))
+        if uri:
+            self.uri = uri
+        else:
+            self.uri = QgsDataSourceURI()
+            if service:
+                self.uri.setConnection(service, dbname, user, passwd)
+            else:
+                self.uri.setConnection(host, port, dbname, user, passwd)
+
+        conninfo = self.uri.connectionInfo(False)
+        err = None
+        for i in range(4):
+            try:
+                self.con = psycopg2.connect(self.uri.connectionInfo(True).encode('utf-8'))
+                if err is not None:
+                    QgsCredentials.instance().put(conninfo,
+                                                  self.uri.username(),
+                                                  self.uri.password())
+                break
+            except psycopg2.OperationalError as e:
+                if i == 3:
+                    raise DbError(unicode(e))
+
+                err = unicode(e)
+                user = self.uri.username()
+                password = self.uri.password()
+                (ok, user, password) = QgsCredentials.instance().get(conninfo,
+                                                                     user,
+                                                                     password,
+                                                                     err)
+                if not ok:
+                    raise DbError(u'Action cancelled by user')
+                if user:
+                    self.uri.setUsername(user)
+                if password:
+                    self.uri.setPassword(password)
+            finally:
+                # remove certs (if any) of the expanded connectionInfo
+                expandedUri = QgsDataSourceURI(self.uri)
```

Comment:
you forgot to create uri from an expanded uri, self.uri seems not expanded => certs will be not removed